File: backend/app/core/sentry.py
"""Sentry error tracking configuration for production"""

import sentry_sdk
from sentry_sdk.integrations.fastapi import FastApiIntegration
from sentry_sdk.integrations.sqlalchemy import SqlalchemyIntegration
import os

def init_sentry():
    """Initialize Sentry error tracking and performance monitoring"""
    
    sentry_dsn = os.getenv("SENTRY_DSN", "")
    environment = os.getenv("ENVIRONMENT", "production")
    
    if not sentry_dsn:
        print("⚠️ SENTRY_DSN not configured, error tracking disabled")
        return
    
    try:
        sentry_sdk.init(
            dsn=sentry_dsn,
            environment=environment,
            
            # Performance monitoring (10% of transactions)
            traces_sample_rate=float(os.getenv("SENTRY_TRACES_SAMPLE_RATE", "0.1")),
            
            # Profiling (10% of transactions)
            profiles_sample_rate=0.1,
            
            # Integrations
            integrations=[
                FastApiIntegration(
                    transaction_style="endpoint",
                    failed_request_status_codes=[500, 501, 502, 503, 504]
                ),
                SqlalchemyIntegration(),
            ],
            
            # Release tracking
            release=os.getenv("APP_VERSION", "1.0.0"),
            
            # Filter events before sending
            before_send=before_send_hook,
            
            # Additional options
            attach_stacktrace=True,
            send_default_pii=False,  # Don't send PII
        )
        
        print(f"✅ Sentry initialized (env: {environment}, DSN: {sentry_dsn[:30]}...)")
        
    except Exception as e:
        print(f"❌ Failed to initialize Sentry: {e}")

def before_send_hook(event, hint):
    """
    Filter and modify events before sending to Sentry
    """
    
    # Don't send 404 errors
    if event.get("logger") == "uvicorn.error":
        if "404" in str(event.get("message", "")):
            return None
    
    # Don't send health check errors
    if "/health" in str(event.get("request", {}).get("url", "")):
        return None
    
    # Add custom tags
    event.setdefault("tags", {})
    event["tags"]["app"] = "iliski-analiz"
    event["tags"]["component"] = "backend"
    
    # Add user context (if available, without PII)
    # Note: Don't add email or name, just user ID
    if "user" in event:
        user_data = event["user"]
        if "email" in user_data:
            del user_data["email"]  # Remove PII
        if "username" in user_data:
            del user_data["username"]  # Remove PII
    
    return event

def capture_exception(exception: Exception, context: dict = None):
    """
    Manually capture an exception with additional context
    
    Usage:
        try:
            risky_operation()
        except Exception as e:
            capture_exception(e, {"user_id": user_id, "operation": "analysis"})
    """
    if context:
        with sentry_sdk.push_scope() as scope:
            for key, value in context.items():
                scope.set_context(key, value)
            sentry_sdk.capture_exception(exception)
    else:
        sentry_sdk.capture_exception(exception)

def capture_message(message: str, level: str = "info", context: dict = None):
    """
    Capture a message (not an error) with context
    
    Usage:
        capture_message(
            "User completed onboarding",
            level="info",
            context={"user_id": user_id}
        )
    """
    if context:
        with sentry_sdk.push_scope() as scope:
            for key, value in context.items():
                scope.set_tag(key, str(value))
            sentry_sdk.capture_message(message, level=level)
    else:
        sentry_sdk.capture_message(message, level=level)

File: frontend/lib/sentry.ts
/**
 * Sentry Error Tracking for Frontend
 *
 * Initializes Sentry for error monitoring and performance tracking.
 */

import * as Sentry from "@sentry/nextjs";
import { env, isProduction } from "./env";

export function initSentry(): void {
  if (!env.NEXT_PUBLIC_SENTRY_DSN) {
    console.warn("⚠️ SENTRY_DSN not configured, error tracking disabled");
    return;
  }

  try {
    Sentry.init({
      dsn: env.NEXT_PUBLIC_SENTRY_DSN,
      environment: env.NEXT_PUBLIC_SENTRY_ENVIRONMENT,

      // Performance Monitoring
      tracesSampleRate: isProduction ? 0.1 : 1.0, // 10% in production, 100% in dev

      // Session Replay (optional)
      replaysSessionSampleRate: 0.1, // 10% of sessions
      replaysOnErrorSampleRate: 1.0, // 100% of sessions with errors

      // Release tracking
      release: env.NEXT_PUBLIC_APP_VERSION,

      // Trace propagation targets for performance monitoring
      tracePropagationTargets: ["localhost", env.NEXT_PUBLIC_API_URL, /^\//],

      // Error filtering
      beforeSend(event, hint) {
        // Don't send errors in development
        if (!isProduction) {
          console.error("Sentry Event (dev):", event, hint);
          return null;
        }

        // Filter out specific errors
        if (event.exception) {
          const error = hint.originalException;
          if (error instanceof Error) {
            // Don't send network errors
            if (error.message.includes("NetworkError")) {
              return null;
            }
            // Don't send aborted requests
            if (error.message.includes("AbortError")) {
              return null;
            }
          }
        }

        // Add custom tags
        event.tags = {
          ...event.tags,
          app: "iliski-analiz",
          component: "frontend",
        };

        return event;
      },
    });

    console.log(
      `✅ Sentry initialized (env: ${env.NEXT_PUBLIC_SENTRY_ENVIRONMENT})`,
    );
  } catch (error) {
    console.error("❌ Failed to initialize Sentry:", error);
  }
}

/**
 * Capture an exception with additional context
 */
export function captureException(
  error: Error,
  context?: Record<string, any>,
): void {
  if (context) {
    Sentry.withScope((scope) => {
      Object.entries(context).forEach(([key, value]) => {
        scope.setContext(key, value);
      });
      Sentry.captureException(error);
    });
  } else {
    Sentry.captureException(error);
  }
}

/**
 * Capture a message (not an error)
 */
export function captureMessage(
  message: string,
  level: Sentry.SeverityLevel = "info",
  context?: Record<string, any>,
): void {
  if (context) {
    Sentry.withScope((scope) => {
      Object.entries(context).forEach(([key, value]) => {
        scope.setTag(key, String(value));
      });
      Sentry.captureMessage(message, level);
    });
  } else {
    Sentry.captureMessage(message, level);
  }
}

/**
 * Set user context for error tracking
 */
export function setUser(user: { id: string; email?: string }): void {
  Sentry.setUser({
    id: user.id,
    // Don't send email in production for privacy
    ...(isProduction ? {} : { email: user.email }),
  });
}

/**
 * Clear user context
 */
export function clearUser(): void {
  Sentry.setUser(null);
}
